File: Modules/MeasureObjectIntensity.m
function handles = MeasureObjectIntensity(handles,varargin)

% Help for the Measure Object Intensity module:
% Category: Measurement
%
% SHORT DESCRIPTION:
% Measures several intensity features for identified objects.
% *************************************************************************
%
% Given an image with objects identified (e.g. nuclei or cells), this
% module extracts intensity features for each object based on a
% corresponding grayscale image. Measurements are recorded for each object.
%
% Features measured:       Feature Number:
% IntegratedIntensity     |       1
% MeanIntensity           |       2
% StdIntensity            |       3
% MinIntensity            |       4
% MaxIntensity            |       5
% IntegratedIntensityEdge |       6
% MeanIntensityEdge       |       7
% StdIntensityEdge        |       8
% MinIntensityEdge        |       9
% MaxIntensityEdge        |      10
% MassDisplacement        |      11
% LowerQuartileIntensity  |      12
% MedianIntensity         |      13
% UpperQuartileIntensity  |      14
%
% How it works:
% Retrieves objects in label matrix format and a corresponding original
% grayscale image and makes measurements of the objects. The label matrix
% image should be "compacted": that is, each number should correspond to an
% object, with no numbers skipped. So, if some objects were discarded from
% the label matrix image, the image should be converted to binary and
% re-made into a label matrix image before feeding it to this module.
%
% Intensity Measurement descriptions:
%
% * IntegratedIntensity - The sum of the pixel intensities within an
% object.
% * MeanIntensity - The average pixel intensity within an object.
% * StdIntensity - The standard deviation of the pixel intensities within
% an object.
% * MaxIntensity - The maximal pixel intensity within an object.
% * MinIntensity - The minimal pixel intensity within an object.
% * IntegratedIntensityEdge - The sum of the edge pixel intensities of an
% object.
% * MeanIntensityEdge - The average edge pixel intensity of an object.
% * StdIntensityEdge - The standard deviation of the edge pixel intensities
% of an object.
% * MaxIntensityEdge - The maximal edge pixel intensity of an object.
% * MinIntensityEdge - The minimal edge pixel intensity of an object.
% * MassDisplacement - The distance between the centers of gravity in the
% gray-level representation of the object and the binary representation of
% the object.
% * LowerQuartileIntensity - the intensity value of the pixel for which 25%
% of the pixels in the object have lower values.
% * MedianIntensity - the median intensity value within the object
% * UpperQuartileIntensity - the intensity value of the pixel for which 75%
% of the pixels in the object have lower values.
%
% For publication purposes, it is important to note that the units of
% intensity from microscopy images are usually described as "Intensity
% units" or "Arbitrary intensity units" since microscopes are not 
% callibrated to an absolute scale. Also, it is important to note whether 
% you are reporting either the mean or the integrated intensity, so specify
% "Mean intensity units" or "Integrated intensity units" accordingly.
%
% See also MeasureImageIntensity.

% CellProfiler is distributed under the GNU General Public License.
% See the accompanying file LICENSE for details.
%
% Developed by the Whitehead Institute for Biomedical Research.
% Copyright 2003,2004,2005.
%
% Please see the AUTHORS file for credits.
%
% Website: http://www.cellprofiler.org
%
% $Revision$

%%%%%%%%%%%%%%%%%
%%% VARIABLES %%%
%%%%%%%%%%%%%%%%%
drawnow

[CurrentModule, CurrentModuleNum, ModuleName] = CPwhichmodule(handles);

%textVAR01 = What did you call the greyscale images you want to measure?
%infotypeVAR01 = imagegroup
ImageName = char(handles.Settings.VariableValues{CurrentModuleNum,1});
%inputtypeVAR01 = popupmenu

%textVAR02 = What did you call the objects that you want to measure?
%choiceVAR02 = Do not use
%infotypeVAR02 = objectgroup
ObjectNameList{1} = char(handles.Settings.VariableValues{CurrentModuleNum,2});
%inputtypeVAR02 = popupmenu

%textVAR03 =
%choiceVAR03 = Do not use
%infotypeVAR03 = objectgroup
ObjectNameList{2} = char(handles.Settings.VariableValues{CurrentModuleNum,3});
%inputtypeVAR03 = popupmenu

%textVAR04 =
%choiceVAR04 = Do not use
%infotypeVAR04 = objectgroup
ObjectNameList{3} = char(handles.Settings.VariableValues{CurrentModuleNum,4});
%inputtypeVAR04 = popupmenu

%textVAR05 =
%choiceVAR05 = Do not use
%infotypeVAR05 = objectgroup
ObjectNameList{4} = char(handles.Settings.VariableValues{CurrentModuleNum,5});
%inputtypeVAR05 = popupmenu

%textVAR06 =
%choiceVAR06 = Do not use
%infotypeVAR06 = objectgroup
ObjectNameList{5} = char(handles.Settings.VariableValues{CurrentModuleNum,6});
%inputtypeVAR06 = popupmenu

%textVAR07 =
%choiceVAR07 = Do not use
%infotypeVAR07 = objectgroup
ObjectNameList{6} = char(handles.Settings.VariableValues{CurrentModuleNum,7});
%inputtypeVAR07 = popupmenu

%%% Initialize measurement structure
BasicFeatures    = {'IntegratedIntensity',...
    'MeanIntensity',...
    'StdIntensity',...
    'MinIntensity',...
    'MaxIntensity',...
    'IntegratedIntensityEdge',...
    'MeanIntensityEdge',...
    'StdIntensityEdge',...
    'MinIntensityEdge',...
    'MaxIntensityEdge',...
    'MassDisplacement',...
    'LowerQuartileIntensity',...
    'MedianIntensity',...
    'UpperQuartileIntensity'};

%%%%%%%%%%%%%%%%
%%% FEATURES %%%
%%%%%%%%%%%%%%%%

if nargin > 1 
    switch varargin{1}
%feature:categories
        case 'categories'
            if nargin == 1 || ismember(varargin{2},ObjectNameList)
                result = { 'Intensity' };
            else
                result = {};
            end
%feature:measurements
        case 'measurements'
            result = {};
            if nargin >= 3 &&...
                strcmp(varargin{3},'Intensity') &&...
                ismember(varargin{2},ObjectNameList)
                result = BasicFeatures;
            end
        otherwise
            error(['Unhandled category: ',varargin{1}]);
    end
    handles=result;
    return;
end

%%%VariableRevisionNumber = 2

%%% Set up the window for displaying the results
ThisModuleFigureNumber = handles.Current.(['FigureNumberForModule',CurrentModule]);
if any(findobj == ThisModuleFigureNumber);
    CPfigure(handles,'Text',ThisModuleFigureNumber);
    columns = 1;
end

%%% START LOOP THROUGH ALL THE OBJECTS
for i = 1:length(ObjectNameList)
    ObjectName = ObjectNameList{i};
    if strcmpi(ObjectName,'Do not use')
        continue
    end

    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %%% PRELIMINARY CALCULATIONS & FILE HANDLING %%%
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    drawnow

    %%% Reads (opens) the image you want to analyze and assigns it to a variable,
    %%% "OrigImage".
    OrigImage = CPretrieveimage(handles,ImageName,ModuleName);

    %%% If the image is three dimensional (i.e. color), the three channels
    %%% are added together in order to measure object intensity.
    if ndims(OrigImage) ~= 2
        s = size(OrigImage);
        if (length(s) == 3 && s(3) == 3)
            OrigImage = OrigImage(:,:,1)+OrigImage(:,:,2)+OrigImage(:,:,3);
        else
            error(['Image processing was canceled in the ', ModuleName, ' module. There was a problem with the dimensions. The image must be grayscale or RGB color.'])
        end
    end

    %%% Retrieves the label matrix image that contains the segmented objects which
    %%% will be measured with this module.
    LabelMatrixImage = CPretrieveimage(handles,['Segmented', ObjectName],ModuleName,'MustBeGray','DontCheckScale');

    %%% For the cases where the label matrix was produced from a cropped
    %%% image, the sizes of the images will not be equal. So, we crop the
    %%% LabelMatrix and try again to see if the matrices are then the
    %%% proper size. Removes Rows and Columns that are completely blank.
    if any(size(OrigImage) < size(LabelMatrixImage))
        ColumnTotals = sum(LabelMatrixImage,1);
        RowTotals = sum(LabelMatrixImage,2)';
        warning off all
        ColumnsToDelete = ~logical(ColumnTotals);
        RowsToDelete = ~logical(RowTotals);
        warning on all
        drawnow
        CroppedLabelMatrix = LabelMatrixImage;
        CroppedLabelMatrix(:,ColumnsToDelete,:) = [];
        CroppedLabelMatrix(RowsToDelete,:,:) = [];
        clear LabelMatrixImage
        LabelMatrixImage = CroppedLabelMatrix;
        %%% In case the entire image has been cropped away, we store a single
        %%% zero pixel for the variable.
        if isempty(LabelMatrixImage)
            LabelMatrixImage = 0;
        end
    end

    if any(size(OrigImage) ~= size(LabelMatrixImage))
        error(['Image processing was canceled in the ', ModuleName, ' module. The size of the image you want to measure is not the same as the size of the image from which the ',ObjectName,' objects were identified.'])
    end
        
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %%% MAKE MEASUREMENTS & SAVE TO HANDLES STRUCTURE %%%
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    drawnow

    %%% Get pixel indexes (fastest way), and count objects
    [sr sc] = size(LabelMatrixImage);        
    props = regionprops(LabelMatrixImage,'PixelIdxList','Area');
    ObjectCount = length(props);    

    %%% Label-aware boundary finding (even when two objects are adjacent)
    LabelBoundaryImage = CPlabelperim(LabelMatrixImage);
    
    if ObjectCount > 0
        Basic = cell(ObjectCount,length(BasicFeatures));
        
        for Object = 1:ObjectCount
            %%% It's possible for objects not to have any pixels,
            %%% particularly tertiary objects (such as cytoplasm from
            %%% cells the exact same size as their nucleus).
            if isempty(props(Object).PixelIdxList),
                [Basic{Object,:}] = deal(0);
                continue;
            end
         
            %%% Measure basic set of Intensity features
            Basic{Object,1} = sum(OrigImage(props(Object).PixelIdxList));
            Basic{Object,2} = mean(OrigImage(props(Object).PixelIdxList));
            Basic{Object,3} = std(OrigImage(props(Object).PixelIdxList));
            Basic{Object,4} = min(OrigImage(props(Object).PixelIdxList));
            Basic{Object,5} = max(OrigImage(props(Object).PixelIdxList));     
     
            %%% Kyungnam, 2007-Aug-06: optimized code
            %%% Cut patch so that we don't have to deal with entire image
            [r,c] = ind2sub([sr sc],props(Object).PixelIdxList);
            rmax = min(sr,max(r));
            rmin = max(1,min(r));
            cmax = min(sc,max(c));
            cmin = max(1,min(c)); 
            BWim = LabelMatrixImage(rmin:rmax,cmin:cmax) == Object;
            Greyim = OrigImage(rmin:rmax,cmin:cmax);
            Boundaryim = LabelBoundaryImage(rmin:rmax,cmin:cmax) == Object;
            perim = Greyim(Boundaryim(:));
            Basic{Object,6}  = sum(perim);
            Basic{Object,7}  = mean(perim);
            Basic{Object,8}  = std(perim);
            Basic{Object,9}  = min(perim);
            Basic{Object,10} = max(perim);
          
            %%% Kyungnam, 2007-Aug-06: the original old code left commented below
            %%%                        'bwperim' is slow!                         
            %             %%% Get perimeter in order to calculate edge features
            %             perim = bwperim(BWim);
            %             perim = Greyim(find(perim)); %#ok Ignore MLint
            %             Basic(Object,6)  = sum(perim);
            %             Basic(Object,7)  = mean(perim);
            %             Basic(Object,8)  = std(perim);
            %             Basic(Object,9)  = min(perim);
            %             Basic(Object,10) = max(perim);   
           
        end
        %%% Calculate the Mass displacment (taking the pixelsize into account), which is the distance between
        %%% the center of gravity in the gray level image and the binary
        %%% image.
        mask = (LabelMatrixImage > 0);
        masked_labels = LabelMatrixImage(mask);
        masked_intensity = OrigImage(mask);
        [x,y] = meshgrid(1:size(LabelMatrixImage,1),1:size(LabelMatrixImage,2));
        masked_x = x(mask);
        masked_y = y(mask);
        CM_x = full(sparse(masked_labels, 1, masked_x) ./ sparse(masked_labels, 1, 1));
        CM_y = full(sparse(masked_labels, 1, masked_y) ./ sparse(masked_labels, 1, 1));
        
        denom = sparse(masked_labels, 1, masked_intensity);
        if denom ~= 0
            intensity_CM_x = full(sparse(masked_labels, 1, masked_x .*masked_intensity) ./ denom);
            intensity_CM_y = full(sparse(masked_labels, 1, masked_y .*masked_intensity) ./ denom);
        else
            intensity_CM_x = zeros(size(CM_x));
            intensity_CM_y = zeros(size(CM_y));
        end

        PixelSize = str2double(handles.Settings.PixelSize);
        diff_x = CM_x - intensity_CM_x;
        diff_y = CM_y - intensity_CM_y;
        Basic(:,11) = arrayfun(@(x) {x}, sqrt(diff_x.^2+diff_y.^2).*PixelSize);
        %
        % A trick for median, lower & upper quartile:
        %   Add the object # to an intensity scaled between .1 and .9
        %   Sort the resulting array.
        %   Restore the pixels to scaled intensities w/o object #
        %   Do the cumulative sum of the areas of each object
        %   Subtract 1/4, 1/2 and 3/4 of the area and use that to
        %   index into the sorted array to get the values.
        %
        SortedObjectPixels=OrigImage(LabelMatrixImage>0);
        Min = min(SortedObjectPixels);
        Max = max(SortedObjectPixels);
        Scale = (Max-Min) / .8;
        SortedObjectPixels = ((SortedObjectPixels - Min) / (Scale+eps))+.1;
        SortedObjectPixels = SortedObjectPixels + LabelMatrixImage(LabelMatrixImage>0);
        SortedObjectPixels = sort(SortedObjectPixels);
        SortedObjectPixels = SortedObjectPixels - floor(SortedObjectPixels);
        SortedObjectPixels = (SortedObjectPixels - .1) * Scale + Min;
        Address = cumsum([props.Area]);
        idx = max(1,floor(Address-[props.Area]*3/4));
        Basic(:,12) = arrayfun(@(x) {x}, SortedObjectPixels(idx));
        idx = max(1,floor(Address-[props.Area]/2));
        Basic(:,13) = arrayfun(@(x) {x}, SortedObjectPixels(idx));
        idx = max(1,floor(Address-[props.Area]/4));
        Basic(:,14) = arrayfun(@(x) {x}, SortedObjectPixels(idx));
    else
        % Fill in with empty sets
        Basic = cell(1,length(BasicFeatures));
    end
    %%% Save measurements
    for j = 1:size(Basic,2)
        feature_name = CPjoinstrings('Intensity',BasicFeatures{j},ImageName);
        handles = CPaddmeasurements(handles, ObjectName, ...
            feature_name, cat(1,Basic{:,j}));
    end
    
    %%% Report measurements
    if any(findobj == ThisModuleFigureNumber);
        % Remove uicontrols from last cycle
        delete(findobj(ThisModuleFigureNumber,'tag','TextUIControl'));
        
        FontSize = handles.Preferences.FontSize;
        if handles.Current.SetBeingAnalyzed == handles.Current.StartingImageSet
            delete(findobj('parent',ThisModuleFigureNumber,'string','R'));
            delete(findobj('parent',ThisModuleFigureNumber,'string','G'));
            delete(findobj('parent',ThisModuleFigureNumber,'string','B'));
        end
        %%%% This first block writes the same text several times
        %%% Header

        uicontrol(ThisModuleFigureNumber,'style','text','units','normalized', 'position', [0 0.95 1 0.04],...
            'HorizontalAlignment','center','BackgroundColor',[.7 .7 .9],'fontname','Helvetica','tag','TextUIControl',...
            'fontsize',FontSize,'fontweight','bold','string',sprintf(['Average intensity features for ', ImageName,', cycle #%d'],handles.Current.SetBeingAnalyzed));

        %%% Number of objects
        uicontrol(ThisModuleFigureNumber,'style','text','units','normalized', 'position', [0.05 0.85 0.3 0.03],...
            'HorizontalAlignment','left','BackgroundColor',[.7 .7 .9],'fontname','Helvetica','tag','TextUIControl',...
            'fontsize',FontSize,'fontweight','bold','string','Number of objects:');

        %%% Text for Basic features
        uicontrol(ThisModuleFigureNumber,'style','text','units','normalized', 'position', [0.05 0.8 0.3 0.03],...
            'HorizontalAlignment','left','BackgroundColor',[.7 .7 .9],'fontname','Helvetica','tag','TextUIControl',...
            'fontsize',FontSize,'fontweight','bold','string','Intensity feature:');
        for k = 1:length(BasicFeatures)
            uicontrol(ThisModuleFigureNumber,'style','text','units','normalized', 'position', [0.05 0.8-0.04*k 0.3 0.03],...
                'HorizontalAlignment','left','BackgroundColor',[.7 .7 .9],'fontname','Helvetica','tag','TextUIControl',...
                'fontsize',FontSize,'string',BasicFeatures{k});
        end

        %%% The name of the object image
        uicontrol(ThisModuleFigureNumber,'style','text','units','normalized', 'position', [0.35+0.1*(columns-1) 0.9 0.1 0.03],...
            'HorizontalAlignment','center','BackgroundColor',[.7 .7 .9],'fontname','Helvetica','tag','TextEachObjUIEachOControl',...
            'fontsize',FontSize,'fontweight','bold','string',ObjectName);

        %%% Number of objects
        uicontrol(ThisModuleFigureNumber,'style','text','units','normalized', 'position', [0.35+0.1*(columns-1) 0.85 0.1 0.03],...
            'HorizontalAlignment','center','BackgroundColor',[.7 .7 .9],'fontname','Helvetica','tag','TextEachObjUIControl',...
            'fontsize',FontSize,'string',num2str(ObjectCount));

        if ObjectCount > 0
            %%% Basic features
            for k = 1:length(BasicFeatures)
                uicontrol(ThisModuleFigureNumber,'style','text','units','normalized', 'position', [0.35+0.1*(columns-1) 0.8-0.04*k 0.1 0.03],...
                    'HorizontalAlignment','center','BackgroundColor',[.7 .7 .9],'fontname','Helvetica','tag','TextEachObjUIControl',...
                    'fontsize',FontSize,'string',sprintf('%0.2f',mean(cat(1,Basic{:,k}))));
            end
        end
        %%% This variable is used to write results in the correct column
        %%% and to determine the correct window size
        columns = columns + 1;
    end
end

File: pyCellProfiler/cellprofiler/modules/measureobjectintensity.py
"""<b>MeasureObjectIntensity</b> - Measures several intensity features for identified objects.
<hr>
Given an image with objects identified (e.g. nuclei or cells), this
module extracts intensity features for each object based on a
corresponding grayscale image. Measurements are recorded for each object.

Retrieves objects in label matrix format and a corresponding original
grayscale image and makes measurements of the objects. The label matrix
image should be "compacted": that is, each number should correspond to an
object, with no numbers skipped. So, if some objects were discarded from
the label matrix image, the image should be converted to binary and
re-made into a label matrix image before feeding it to this module.

Features that can be measured by this module:
<ul><li><i>IntegratedIntensity:</i> The sum of the pixel intensities within an
 object.</li>
<li><i>MeanIntensity:</i> The average pixel intensity within an object.</li>
<li><i>StdIntensity:</i> The standard deviation of the pixel intensities within
 an object.</li>
<li><i>MaxIntensity:</i> The maximal pixel intensity within an object.</li>
<li><i>MinIntensity:</i> The minimal pixel intensity within an object.</li>
<li><i>IntegratedIntensityEdge:</i> The sum of the edge pixel intensities of an
 object.</li>
<li><i>MeanIntensityEdge:</i> The average edge pixel intensity of an object.</li>
<li><i>StdIntensityEdge:</i> The standard deviation of the edge pixel intensities
 of an object.</li>
<li><i>MaxIntensityEdge:</i> The maximal edge pixel intensity of an object.</li>
<li><i>MinIntensityEdge:</i> The minimal edge pixel intensity of an object.</li>
<li><i>MassDisplacement:</i> The distance between the centers of gravity in the
 gray-level representation of the object and the binary representation of
 the object.</li>
<li><i>LowerQuartileIntensity:</i> The intensity value of the pixel for which 25%
 of the pixels in the object have lower values.</li>
<li><i>MedianIntensity:</i> The median intensity value within the object</li>
<li><i>UpperQuartileIntensity:</i> The intensity value of the pixel for which 75%
 of the pixels in the object have lower values.</li></ul>

For publication purposes, it is important to note that the units of
intensity from microscopy images are usually described as "Intensity
units" or "Arbitrary intensity units" since microscopes are not 
callibrated to an absolute scale. Also, it is important to note whether 
you are reporting either the mean or the integrated intensity, so specify
"Mean intensity units" or "Integrated intensity units" accordingly.

See also <b>MeasureImageIntensity</b>.
"""
#CellProfiler is distributed under the GNU General Public License.
#See the accompanying file LICENSE for details.
#
#Developed by the Broad Institute
#Copyright 2003-2009
#
#Please see the AUTHORS file for credits.
#
#Website: http://www.cellprofiler.org

__version__="$Revision$"

import numpy as np
import scipy.ndimage as nd

import cellprofiler.cpmodule as cpm
import cellprofiler.measurements as cpmeas
import cellprofiler.settings as cps
import cellprofiler.cpmath.outline as cpmo
from cellprofiler.cpmath.cpmorphology import fixup_scipy_ndimage_result as fix
from cellprofiler.cpmath.filter import stretch

INTENSITY = 'Intensity'
INTEGRATED_INTENSITY = 'IntegratedIntensity'
MEAN_INTENSITY = 'MeanIntensity'
STD_INTENSITY = 'StdIntensity'
MIN_INTENSITY = 'MinIntensity'
MAX_INTENSITY = 'MaxIntensity'
INTEGRATED_INTENSITY_EDGE = 'IntegratedIntensityEdge'
MEAN_INTENSITY_EDGE = 'MeanIntensityEdge'
STD_INTENSITY_EDGE = 'StdIntensityEdge'
MIN_INTENSITY_EDGE = 'MinIntensityEdge'
MAX_INTENSITY_EDGE = 'MaxIntensityEdge'
MASS_DISPLACEMENT = 'MassDisplacement'
LOWER_QUARTILE_INTENSITY = 'LowerQuartileIntensity'
MEDIAN_INTENSITY = 'MedianIntensity'
UPPER_QUARTILE_INTENSITY = 'UpperQuartileIntensity'

ALL_MEASUREMENTS = [INTEGRATED_INTENSITY, MEAN_INTENSITY, STD_INTENSITY,
                        MIN_INTENSITY, MAX_INTENSITY, INTEGRATED_INTENSITY_EDGE,
                        MEAN_INTENSITY_EDGE, STD_INTENSITY_EDGE,
                        MIN_INTENSITY_EDGE, MAX_INTENSITY_EDGE, 
                        MASS_DISPLACEMENT, LOWER_QUARTILE_INTENSITY,
                        MEDIAN_INTENSITY, UPPER_QUARTILE_INTENSITY]

class MeasureObjectIntensity(cpm.CPModule):

    module_name = "MeasureObjectIntensity"
    variable_revision_number = 3
    category = "Measurement"
    
    def create_settings(self):
        self.images = []
        self.add_image()
        self.image_count = cps.HiddenCount(self.images)
        self.add_image_button = cps.DoSomething("", "Add image", self.add_image)
        self.divider = cps.Divider()
        self.objects = []
        self.add_object()
        self.add_object_button = cps.DoSomething("", "Add object", self.add_object)

    def add_image(self):
        group = cps.SettingsGroup()
        group.append("name", cps.ImageNameSubscriber("Select input image:","None", doc = 
                                                     """What did you call the grayscale images you want to process?"""))
        group.append("remover", cps.RemoveSettingButton("", "Remove above image", self.images, group))
        self.images.append(group)

    def add_object(self):
        group = cps.SettingsGroup()
        group.append("name", cps.ObjectNameSubscriber("Select objects to measure:","None", doc = 
                                                          """What did you call the objects that you want to measure?"""))
        group.append("remover", cps.RemoveSettingButton("", "Remove above image", self.images, group))
        self.objects.append(group)

    def settings(self):
        result = [self.image_count]
        result += [im.name for im in self.images]
        result += [obj.name for obj in self.objects]
        return result

    def visible_settings(self):
        result = []
        for im in self.images:
            result += im.unpack_group()
        result += [self.add_image_button, self.divider]
        for im in self.objects:
            result += im.unpack_group()
        result += [self.add_object_button]
        return result
        
    def upgrade_settings(self,setting_values,variable_revision_number,
                         module_name,from_matlab):
        '''Adjust setting values if they came from a previous revision
        
        setting_values - a sequence of strings representing the settings
                         for the module as stored in the pipeline
        variable_revision_number - the variable revision number of the
                         module at the time the pipeline was saved. Use this
                         to determine how the incoming setting values map
                         to those of the current module version.
        module_name - the name of the module that did the saving. This can be
                      used to import the settings from another module if
                      that module was merged into the current module
        from_matlab - True if the settings came from a Matlab pipeline, False
                      if the settings are from a CellProfiler 2.0 pipeline.
        
        Overriding modules should return a tuple of setting_values,
        variable_revision_number and True if upgraded to CP 2.0, otherwise
        they should leave things as-is so that the caller can report
        an error.
        '''
        if from_matlab and variable_revision_number == 2:
            # Old matlab-style. Erase any setting values that are
            # "Do not use"
            new_setting_values = [setting_values[0],cps.DO_NOT_USE]
            for setting_value in setting_values[1:]:
                if setting_value != cps.DO_NOT_USE:
                    new_setting_values.append(setting_value)
            setting_values = new_setting_values
            from_matlab = False
            variable_revision_number = 2
        if variable_revision_number == 2:
            assert not from_matlab
            num_imgs = setting_values.index(cps.DO_NOT_USE)
            setting_values = [str(num_imgs)] + setting_values[:num_imgs] + setting_values[num_imgs+1:]
            variable_revision_number = 3
        return setting_values, variable_revision_number, from_matlab

    def prepare_settings(self,setting_values):
        """Do any sort of adjustment to the settings required for the given values
        
        setting_values - the values for the settings just prior to mapping
                         as done by set_settings_from_values
        This method allows a module to specialize itself according to
        the number of settings and their value. For instance, a module that
        takes a variable number of images or objects can increase or decrease
        the number of relevant settings so they map correctly to the values.
        
        See cellprofiler.modules.measureobjectareashape for an example.
        """
        #
        # The settings have two parts - images, then objects
        # The parts are divided by the string, cps.DO_NOT_USE
        #
        image_count = int(setting_values[0])
        object_count = len(setting_values) - image_count - 1
        del self.images[image_count:]
        while len(self.images) < image_count:
            self.add_image()
        del self.objects[object_count:]
        while len(self.objects) < object_count:
            self.add_object()

    def get_measurement_columns(self, pipeline):
        '''Return the column definitions for measurements made by this module'''
        columns = []
        for image_name in [im.name for im in self.images]:
            for object_name in [obj.name for obj in self.objects]:
                for feature in (ALL_MEASUREMENTS):
                    columns.append((object_name.value,
                                    "%s_%s_%s"%(INTENSITY, feature,
                                                image_name.value),
                                    cpmeas.COLTYPE_FLOAT))
        return columns
            
    def get_categories(self,pipeline, object_name):
        """Get the categories of measurements supplied for the given object name
        
        pipeline - pipeline being run
        object_name - name of labels in question (or 'Images')
        returns a list of category names
        """
        for object_name_variable in [obj.name for obj in self.objects]:
            if object_name_variable.value == object_name:
                return [INTENSITY]
        return []
    
    def get_measurements(self, pipeline, object_name, category):
        """Get the measurements made on the given object in the given category"""
        if category != INTENSITY:
            return []
        for object_name_variable in [obj.name for obj in self.objects]:
            if object_name_variable.value == object_name:
                return ALL_MEASUREMENTS
        return []
    
    def get_measurement_images(self, pipeline,object_name, category, measurement):
        """Get the images used to make the given measurement in the given category on the given object"""
        if category != INTENSITY:
            return []
        if measurement not in ALL_MEASUREMENTS:
            return []
        for object_name_variable in [obj.name for obj in self.objects]:
            if object_name_variable == object_name:
                return [image.name.value for image in self.images]
        return []
    
    def is_interactive(self):
        return False
    
    def run(self, workspace):
        if workspace.frame is not None:
            statistics = [("Image","Object","Feature","Mean","Median","STD")]
            workspace.display_data.statistics = statistics
        for image_name in [img.name for img in self.images]:
            image = workspace.image_set.get_image(image_name.value,
                                                  must_be_grayscale=True)
            img = image.pixel_data
            if image.has_mask:
                masked_image = img.copy()
                masked_image[image.mask] = 0
                img = img[image.mask]
            else:
                masked_image = img
            for object_name in [obj.name for obj in self.objects]:
                objects = workspace.object_set.get_objects(object_name.value)
                labels   = objects.segmented
                nobjects = np.int32(np.max(labels))
                outlines = cpmo.outline(labels)
                
                if image.has_mask:
                    masked_labels = labels.copy()
                    masked_labels[image.mask] = 0
                    labels = labels[image.mask]
                    masked_outlines = outlines.copy()
                    masked_outlines[image.mask] = 0
                    outlines = outlines[image.mask]
                else:
                    masked_labels = labels
                    masked_outlines = outlines
                
                if nobjects > 0:
                    lindexes = np.arange(nobjects)+1
                    integrated_intensity = fix(nd.sum(img, labels, lindexes))
                    integrated_intensity_edge = fix(nd.sum(img, outlines,
                                                           lindexes))
                    mean_intensity = fix(nd.mean(img, labels, lindexes))
                    mean_intensity_edge = fix(nd.mean(img, outlines, lindexes))
                    std_intensity = fix(nd.standard_deviation(img, labels, 
                                                              lindexes))
                    std_intensity_edge = fix(nd.standard_deviation(img, outlines, 
                                                                   lindexes))
                    min_intensity = fix(nd.minimum(img, labels, lindexes))
                    min_intensity_edge = fix(nd.minimum(img, outlines,
                                                        lindexes))
                    max_intensity = fix(nd.maximum(img, labels, lindexes))
                    max_intensity_edge = fix(nd.maximum(img, outlines,
                                                        lindexes))
                else:
                    integrated_intensity = np.zeros((0,))
                    integrated_intensity_edge = np.zeros((0,))
                    mean_intensity = np.zeros((0,))
                    mean_intensity_edge = np.zeros((0,))
                    std_intensity = np.zeros((0,))
                    std_intensity_edge = np.zeros((0,))
                    min_intensity = np.zeros((0,))
                    min_intensity_edge = np.zeros((0,))
                    max_intensity = np.zeros((0,))
                    max_intensity_edge = np.zeros((0,))
                    
                # The mass displacement is the distance between the center
                # of mass of the binary image and of the intensity image. The
                # center of mass is the average X or Y for the binary image
                # and the sum of X or Y * intensity / integrated intensity
                if nobjects > 0:
                    mesh_x, mesh_y = np.meshgrid(range(masked_image.shape[1]),
                                                 range(masked_image.shape[0]))
                    cm_x = fix(nd.mean(mesh_x, masked_labels, lindexes))
                    cm_y = fix(nd.mean(mesh_y, masked_labels, lindexes))
                    
                    i_x = fix(nd.sum(mesh_x * masked_image,masked_labels,
                                          lindexes))
                    i_y = fix(nd.sum(mesh_y * masked_image,masked_labels,
                                          lindexes))
                    cmi_x = i_x / integrated_intensity
                    cmi_y = i_y / integrated_intensity
                    diff_x = cm_x - cmi_x
                    diff_y = cm_y - cmi_y
                    mass_displacement = np.sqrt(diff_x * diff_x+diff_y*diff_y)
                else:
                    mass_displacement = np.zeros((0,))
                
                # We do the quantile measurements using an indexing trick:
                # given a label integer L and the intensity at that label, I
                # L+I is a number between L and L+1. So if you add the label
                # matrix to the intensity matrix and sort, you'll get the
                # intensities in order by label, then by magnitude. If you
                # do a cumsum of areas of labels, you'll get indices into
                # the ordered array and you can read out quantiles pretty easily
                
                if nobjects > 0:
                    stretched_img = stretch(img) * .99
                    flat_img = img.flatten()
                    areas = fix(nd.sum(np.ones(labels.shape,int),
                                       labels, np.arange(nobjects+1)))
                    areas = areas.astype(int)
                    indices = np.cumsum(areas)[:-1]
                    ordered_image = stretched_img + labels.astype(float)
                    ordered_image = ordered_image.flatten()
                    image_idx = np.argsort(ordered_image)
                    max_indices = (indices + areas[1:] - 1).astype(int)
                    indices_25  = indices+(areas[1:]+2)/4
                    indices_50  = indices+(areas[1:]+1)/2
                    indices_75  = indices+3*(areas[1:]+2) / 4
                    #
                    # Check for round-up overflow
                    #
                    for indices in (indices_25, indices_50, indices_75):
                        imask = indices > max_indices
                        indices[imask] = max_indices[imask]
                    lower_quartile_intensity = flat_img[image_idx[indices_25]]
                    median_intensity         = flat_img[image_idx[indices_50]]
                    upper_quartile_intensity = flat_img[image_idx[indices_75]]
                else:
                    lower_quartile_intensity = np.zeros((0,))
                    median_intensity = np.zeros((0,))
                    upper_quartile_intensity = np.zeros((0,))
                
                m = workspace.measurements
                for feature_name, measurement in \
                    ((INTEGRATED_INTENSITY, integrated_intensity),
                     (MEAN_INTENSITY, mean_intensity),
                     (STD_INTENSITY, std_intensity),
                     (MIN_INTENSITY, min_intensity),
                     (MAX_INTENSITY, max_intensity),
                     (INTEGRATED_INTENSITY_EDGE, integrated_intensity_edge),
                     (MEAN_INTENSITY_EDGE, mean_intensity_edge),
                     (STD_INTENSITY_EDGE, std_intensity_edge),
                     (MIN_INTENSITY_EDGE, min_intensity_edge),
                     (MAX_INTENSITY_EDGE, max_intensity_edge),
                     (MASS_DISPLACEMENT, mass_displacement),
                     (LOWER_QUARTILE_INTENSITY, lower_quartile_intensity),
                     (MEDIAN_INTENSITY, median_intensity),
                     (UPPER_QUARTILE_INTENSITY, upper_quartile_intensity)):
                    measurement_name = "%s_%s_%s"%(INTENSITY,feature_name,
                                                   image_name.value)
                    m.add_measurement(object_name.value,measurement_name, 
                                      measurement)
                    if workspace.frame is not None and len(measurement) > 0:
                        statistics.append((image_name.value, object_name.value, 
                                           feature_name,
                                           np.round(np.mean(measurement),3),
                                           np.round(np.median(measurement),3),
                                           np.round(np.std(measurement),3)))
        
    def display(self, workspace):
        figure = workspace.create_or_find_figure(subplots=(1,1))
        figure.subplot_table(0,0,workspace.display_data.statistics,
                             ratio=(.2,.2,.3,.1,.1,.1))
